jobclient/src/main/java/com/twosigma/cook/jobclient/Constraint.java

```diff
@@ -0,0 +1,170 @@
+        }
+    }
+
+    private String _attribute;
+    private Operator _operator;
+    private String _value;
+
+    Constraint(String attribute, Operator operator, String value) {
+        Preconditions.checkNotNull(attribute);
+        Preconditions.checkNotNull(operator);
+        Preconditions.checkNotNull(value);
+        _attribute = attribute;
+        _operator = operator;
+        _value = value;
+    }
+
+    Constraint(String attribute, String operator, String value) {
+        Preconditions.checkNotNull(attribute);
+        Preconditions.checkNotNull(operator);
+        Preconditions.checkNotNull(value);
+        _attribute = attribute;
+        _operator = Operator.fromString(operator);
+        _value = value;
+    }
+
+    /**
+     * Construct a constraint from a {@link JSONArray}.
+     *
+     * The given {@link JSONArray} is assumed to have three strings that
+     * represent attribute, operator, and value of a constraint respectively.
+     *
+     * @param constraint The constraint in {@link JSONArray} representation.
+     */
+    Constraint(JSONArray constraint) {
+        Preconditions.checkNotNull(constraint);
+        Preconditions.checkArgument(constraint.length() == 3);
+        _attribute = constraint.getString(0);
+        _operator = Operator.fromString(constraint.getString(1));
+        _value = constraint.getString(2);
+    }
+
+    /**
+     * @return the attribute of this constraint.
+     */
+    public String getAttribute() {
+        return _attribute;
+    }
+
+    /**
+     * @return the operator of this constraint.
+     */
+    public Operator getOperator() {
+        return _operator;
+    }
+
+    /**
+     * @return the value of this constraint.
+     */
+    public String getValue() {
+        return _value;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+
+        Constraint that = (Constraint) o;
+
+        if (_attribute != null ? !_attribute.equals(that._attribute) : that._attribute != null) return false;
+        if (_operator != that._operator) return false;
+        return _value != null ? _value.equals(that._value) : that._value == null;
+    }
+
+    @Override
+    public int hashCode() {
+        int result = _attribute != null ? _attribute.hashCode() : 0;
+        result = 31 * result + (_operator != null ? _operator.hashCode() : 0);
+        result = 31 * result + (_value != null ? _value.hashCode() : 0);
+        return result;
+    }
+
+    /**
+     * @return this constraint as a JSONArray.
+     */
+    JSONArray toJson() {
+        JSONArray jsonArray = new JSONArray();
+        jsonArray.put(0, _attribute);
+        jsonArray.put(1, _operator.toString());
+        jsonArray.put(2, _value);
+        return jsonArray;
+    }
+
+    // Supported operators
+
+    final public static Operator EQUALS = Operator.EQUALS;
```

Comment:
Can you move this binding as close to the operator class definition as possible?

Reply:
sure thing